go-controller/hybrid-overlay/pkg/controller/master.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	// Allocate a new host subnet for this node
+	hostsubnets, err := m.allocator.AllocateNetworks()
+	if err != nil {
+		return nil, fmt.Errorf("Error allocating hybrid overlay HostSubnet for node %s: %v", node.Name, err)
+	}
+
+	if err := annotator.Set(types.HybridOverlayNodeSubnet, hostsubnets[0].String()); err != nil {
+		_ = m.allocator.ReleaseNetwork(hostsubnets[0])
+		return nil, err
+	}
+
+	klog.Infof("Allocated hybrid overlay HostSubnet %s for node %s", hostsubnets[0], node.Name)
+	return hostsubnets[0], nil
+}
+
+func (m *MasterController) releaseNodeSubnet(nodeName string, subnet *net.IPNet) error {
+	if err := m.allocator.ReleaseNetwork(subnet); err != nil {
+		return fmt.Errorf("Error deleting hybrid overlay HostSubnet %s for node %q: %s", subnet, nodeName, err)
+	}
+	klog.Infof("Deleted hybrid overlay HostSubnet %s for node %s", subnet, nodeName)
+	return nil
+}
+
+func (m *MasterController) handleOverlayPort(node *kapi.Node, annotator kube.Annotator) error {
+	_, haveDRMACAnnotation := node.Annotations[types.HybridOverlayDRMAC]
+
+	subnet, err := ovn.ParseNodeHostSubnet(node)
+	if subnet == nil || err != nil {
+		// No subnet allocated yet; clean up
+		if haveDRMACAnnotation {
+			m.deleteOverlayPort(node)
+			annotator.Delete(types.HybridOverlayDRMAC)
+		}
+		return nil
+	}
+
+	if haveDRMACAnnotation {
+		// already set up; do nothing
+		return nil
+	}
+
+	portName := houtil.GetHybridOverlayPortName(node.Name)
+	portMAC, portIP, _ := util.GetPortAddresses(portName)
+	if portMAC == nil || portIP == nil {
+		if portIP == nil {
+			// Get the 3rd address in the node's subnet; the first is taken
+			// by the k8s-cluster-router port, the second by the management port
+			first := util.NextIP(subnet.IP)
+			second := util.NextIP(first)
+			portIP = util.NextIP(second)
+		}
+		if portMAC == nil {
+			portMAC, _ = net.ParseMAC(util.IPAddrToHWAddr(portIP))
+		}
+
+		klog.Infof("creating node %s hybrid overlay port", node.Name)
+
+		var stderr string
+		_, stderr, err = util.RunOVNNbctl("--", "--may-exist", "lsp-add", node.Name, portName,
+			"--", "lsp-set-addresses", portName, portMAC.String()+" "+portIP.String())
+		if err != nil {
+			return fmt.Errorf("failed to add hybrid overlay port for node %s"+
+				", stderr:%s: %v", node.Name, stderr, err)
+		}
+
+		if err := util.UpdateNodeSwitchExcludeIPs(node.Name, subnet); err != nil {
+			return err
+		}
+	}
+	if err := annotator.Set(types.HybridOverlayDRMAC, portMAC.String()); err != nil {
+		return fmt.Errorf("failed to set node %s hybrid overlay DRMAC annotation: %v", node.Name, err)
+	}
+
+	return nil
+}
+
+func (m *MasterController) deleteOverlayPort(node *kapi.Node) {
+	klog.Infof("removing node %s hybrid overlay port", node.Name)
+	portName := houtil.GetHybridOverlayPortName(node.Name)
+	_, _, _ = util.RunOVNNbctl("--", "--if-exists", "lsp-del", portName)
+}
+
+// Add handles node additions
+func (m *MasterController) Add(node *kapi.Node) {
+	defer GinkgoRecover()
```

Comment:
ginkgo shouldn't be in here right? This was just from you debugging unit testing.

Reply:
yep, leftover